PR-Review-Guide.md

```diff
@@ -0,0 +1,139 @@
+  - What additional checks would resolve it?
+
+### Reliability
+
+- What happens on failure? (Consider **every** possible failure). For example,
+  - Are temporary files left on the disk?
+  - Are any resources (such as memory and open files) leaked?
+
+- If the code crashes, what is the overall system impact?
+  - Can the system be restarted / recovered?
+  - Could [security](#security) be compromised?
+
+### Performance
+
+- Could this change negatively impact performance?
+- Can any calls block? If so, there should be a log call before the blocking call explaining what is about to be done.
+
+### Maintenance
+
+- How easy would this code be to maintain:
+  - By an experienced developer on the project?
+  - By a developer with none / almost no experience of the code base or the project?
+
+### Usability
+
+- Does the change improve usability or make it worse?
+- Could usability be improved further?
+
+## Specifics to consider
+
+### Basics
+
+- Consider *all* inputs and outputs.
+- What resources are being used? (memory, disk, etc.) Are they released?
+- Are all return values and arguments checked?
+- Are there any magic numbers?
+
+### Layout and formatting
+
+- Is formatting and naming consistent?
+- Are there any spelling mistakes ("typos")?
+- Do all new files contain the required licence header?
+
+### Design
+
+- Could the code be simplified / refactored?
+  - Is the code "too clever"? Overly clever code tends to be fragile: reject it unless there is a **very** good documented reason!
+- Is there a better (simpler) solution to the problem?
+- Is there any duplication?
+- Is the code "re-inventing the wheel?" Is there a standard package that could be used instead?
+- What assumptions does the code make? Are they valid? Are they documented?
+- Does the code make sense?
+  - Could someone with minimal exposure to the codebase guess at what it is doing? (hint: they should be able to).
+
+### Code comments
+
+- Is the code sufficiently well commented?
+- Are all functions and function parameters documented (or for `golang` programs, is their purpose obvious?)
+- Does this change require an update to the `README`(s) / architecture docs?
+- Are any lines commented out? If so, they should be **removed**!
+- Are there any `FIXME`/`TODO`'s? If so, they should be replaced by a full URL to an issue number tracking the work that needs to be done.
+
+### Documentation
+
+- Does the document update need to be applied to any **other** documents?
+  For example, when a PR updates an installation guide, those same changes may need to be applied to one or more of the other installation guides!
+
+- Does the PR make any of the following changes? If so, it may well need a documentation change:
+  - new programs / scripts added or removed.
+  - new program / script options added / removed.
+  - new config options added or removed.
+  - new features added / removed.
+  - existing bug(s) fixed.
+
+### Logging and Debugging
```

Comment:
Use a lowercase d for "debugging" for consistency with the use of sentence case in titles.